Make message control buttons accessible.

Gittenburg opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes #15910.

This should be test deployed on CZO and if no problems are found backported to 3.x.

**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
compose:
![compose_a11y](https://user-images.githubusercontent.com/47354597/89879547-69e5e080-dbc3-11ea-9b7c-516d3ff03453.gif)

edit:
![edit-a11y](https://user-images.githubusercontent.com/47354597/89879541-694d4a00-dbc3-11ea-88e6-0696858a3cd3.gif)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
The first two commits look good.  I think the third is a mistake.  A **lot** of people use the tab + enter to send messages, including users for whom accessibility is a priority; there's no reason to make those users tab through whatever 7 other elements exist in order to use that flow for submitting our "send message" form.

Gittenburg commented:
Oh there absolutely is a reason: Like I explained in the commit message, having the tab order diverge from the DOM order is widely considered to be bad for accessibility:
* https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex:
> Avoid using tabindex values greater than 0. Doing so makes it difficult for people who rely on assistive technology to navigate and operate page content. Instead, write the document with the elements in a logical sequence.
* https://karlgroves.com/2018/11/13/why-using-tabindex-values-greater-than-0-is-bad
* https://developer.paciellogroup.com/blog/2014/08/using-the-tabindex-attribute/
* https://www.w3.org/Bugs/Public/show_bug.cgi?id=27076

If we want to preserve the Tab + Enter shortcut while adhering to accessibility best practices, our only option is to move the Send button to be the first element below the textarea instead of the last.

timabbott commented:
I merged the first 2 commits.  The rest needs more discussion.

andersk commented:
The test deployment on CZO caught a [typeahead-related bug](https://chat.zulip.org/#narrow/stream/9-issues/topic/typeahead.20bug/near/960636) in “compose: Make message control buttons accessible”.

Gittenburg commented:
Fixed.

andersk commented:
The send button is misplaced above the compose box in preview mode.

timabbott commented:
My bigger concern is that I'm not feeling great about the visuals of this change after playing with it for a few days.

timabbott commented:
BTW there's probably this same issue with the "message editing" UI -- I think that's probably fixable with just `tabindex` additions.

Gittenburg commented:
>My bigger concern is that I'm not feeling great about the visuals of this change after playing with it for a few days.

Ok, I updated it to not change the visuals.

>BTW there's probably this same issue with the "message editing" UI 

Fixed in the fourth commit.

Please redeploy on CZO for testing.

timabbott commented:
Can you rebase?

Gittenburg commented:
@timabbott Rebased.

timabbott commented:
I merged the message-edit one as0dbc5df42dfa526c477c9eac9ede3f8fa4ca9985 and opened https://github.com/zulip/zulip/issues/16105 for follow-up improvements that are perhaps their own issue.

timabbott commented on static/styles/compose.scss:
```diff
@@ -349,6 +349,23 @@ input.recipient_box {
 }
 
+#compose-send-button {
+    border-radius: 4px;
```
Maybe worth adding a comment that this floats right in order to ensure the desired tab ordering.

Gittenburg replied:
The `float: right` does not change the tab ordering, just the visual position.

timabbott replied:
I'll add what I have in mind.

timabbott commented on static/js/composebox_typeahead.js:
```diff
@@ -189,78 +185,39 @@ function handle_keydown(e) {
-            // before they fully release the Tab key, the Tab will be lost.  Note that we don't
-            // want to change focus right away in the private_message_recipient box since it
-            // takes the typeaheads a little time to open after the user finishes typing, which
-            // can lead to the focus moving without the autocomplete having a chance to happen.
```
Do you understand whether we're regression behavior here for folks hitting `Tab+Enter` very quickly once this PR is merged?  The old comment here suggests we might be, but it's quite old.

Also, is there an explanation for why we no longer need to check if typeaheads are shown in this logic?

timabbott replied:
I did a round of testing and wasn't able to cause `Tab+Enter` to fail to send in any of my tests, including where I conspicuously held down `Tab` until after I hit `Enter`.

Gittenburg replied:
>Do you understand whether we're regression behavior here for folks hitting Tab+Enter very quickly once this PR is merged? The old comment here suggests we might be, but it's quite old.

I think this was only a problem previously because `nextFocus` was set to `false` in `keydown` under some circumstances (which I no longer do).

>Also, is there an explanation for why we no longer need to check if typeaheads are shown in this logic?

It should never happen that a recipient typeahead is open while the compose textarea has focus and the composebox typeaheads stop the event propagation, so that this code path isn't reached anyway.

timabbott commented:
Redeployed on chat.zulip.org.

timabbott commented:
Merged as the series ending with 2545703b58e65d0bb1e6609866d441b4151538e6, after adding a block comment in the CSS to explain why we're doing this somewhat odd thing.  

Huge thanks for fixing this @Gittenburg!  There's a bit more to do on the accessibility of those buttons in https://github.com/zulip/zulip/issues/16105, but this is a big step forward for Zulip's accessibility.
